src/services/API/RootClient.js

```diff
@@ -77,34 +81,59 @@ client.interceptors.response.use(
-            var newAccessToken = store.state.auth.accessToken.access_token;
-            // Add the new access token to the header of the request
-            error.config.headers["Authorization"] = `Bearer ${newAccessToken}`;
-
+      // reAuthenticate the user if this is the first reAuthentication call. While reAuthentication is going on
+      // if some other request ends up here, then wait for the reAuthentication process to finish. Once that is done,
+      // set the newly retrieved access token as an auth header and retry the request. Any errors in the reauthentication
+      // process are caught and the user is logged out.
+      let run = async () => {
+        try {
+          let reAuthenticationState = store.state.auth.reAuthenticationState;
+          switch (reAuthenticationState) {
+            case "dropped":
+              // re authentication process was dropped, throw an error.
+              throw new Error(
+                "Re-authentication failed, no more calls allowed"
+              );
+            case "in-process":
+              // re authentication process in progress. Wait for the promise to resolve,
+              // then proceed
+              await store.state.auth.reAuthenticationPromise;
+              break;
+            case "not-started":
+              // fresh re authentication process to begin now. Wait for the process to complete,
+              // then proceed
+              await UserFunctionalService.reAuthenticate(store);
+              break;
+          }
+          // Add the new access token recieved from the re authentication process
+          // to the header of the request and retry the request
+          let newAccessToken = store.state.auth.accessToken.access_token;
+          error.config.headers["Authorization"] = `Bearer ${newAccessToken}`;
+          if (error.config.url == userFromTokenEndpoint) {
             // If the call is fetching a user from an access token, we need to update the
             // request params with the new access token
-            if (error.config.url == userFromTokenEndpoint)
-              error.config.params["token"] = newAccessToken;
-
-            // try the request again
-            return client.request(error.config);
+            error.config.params["token"] = newAccessToken;
+          }
+          return client.request(error.config); // retrying the request
+        } catch (err) {
+          // an error occured in the re authentication process
+          // reset the reAuthenticationState value, so any future reAuthentication processes
+          // can function normally
+          store.dispatch("auth/setReAuthenticationState", "not-started");
+          // if the user is still logged in, log them out
+          // if the user is not logged in, then just throw an error
+          if (store.getters["auth/isAuthenticated"]) {
+            store.dispatch("auth/autoLogoutUser");
+            throw new Error("AutoLogout! " + err.message);
```

Comment:
are you using throw now instead of Promise.reject because we are not returning any Promise anymore?

Reply:
`async` functions always return a promise. Always.
Refer to the image below

![image](https://user-images.githubusercontent.com/20169829/140610266-8e0ce650-3668-4410-9ae6-48ea4cc3f18f.png)


So if it was not an async function, I would've written 
```js
return Promise.reject("Autologout")
```

But because the logic is wrapped in an async function, I can just throw an error and it will be wrapped in a promise and that promise will be rejected.

And why are we returning a promise anyway? Turns out we HAVE to return a promise from the interceptor. (you can try returning a non promise and it'll throw an error). We HAVE to return a promise from the interceptor as that is what's being expected by our originator invocations. (the originator invocations are the ones where we make a call like PlioAPISerivce.getPlio())

Reply:
But you cannot return a Promise inside a non-async function, right? So, what does this line mean?
<img width="467" alt="Screenshot 2021-11-08 at 9 37 40 AM" src="https://user-images.githubusercontent.com/13185699/140682840-ad85cff2-9467-489b-9237-87bd7e62875d.png">

Either we should use the `Promise.reject/resolve` thing everything or `return value / throw Error` thing everywhere.  Following both the styles will be confusing for anyone reading the code.